File: src/main/java/io/thegamingmahi/luacord/environment/wrappers/storage/YamlStorage.java
package io.thegamingmahi.luacord.environment.wrappers.storage;

import io.thegamingmahi.luacord.Utilities;
import io.thegamingmahi.luacord.environment.exception.StorageObjectException;
import org.bukkit.configuration.InvalidConfigurationException;
import org.bukkit.configuration.file.YamlConfiguration;
import org.luaj.vm2.LuaBoolean;
import org.luaj.vm2.LuaString;
import org.luaj.vm2.LuaValue;
import org.luaj.vm2.lib.jse.CoerceJavaToLua;

import java.io.FileReader;
import java.io.IOException;

/**
 * @author jammehcow
 */

public class YamlStorage extends StorageObject {
    private YamlConfiguration yamlConfiguration;

    public YamlStorage(LukkitPlugin plugin, String path) {
        super(plugin, path, Storage.YAML);
        this.yamlConfiguration = new YamlConfiguration();

        try {
            this.yamlConfiguration.load(new FileReader(this.getStorageFile()));
        } catch (IOException | InvalidConfigurationException e) {
            e.printStackTrace();
        }
    }

    @Override
    protected LuaBoolean exists(String path) {
        return yamlConfiguration.get(path) != null ? LuaValue.TRUE : LuaValue.FALSE;
    }

    @Override
    public LuaBoolean setDefaultValue(LuaString path, LuaValue value) throws StorageObjectException {
        if (this.yamlConfiguration.get(path.checkjstring()) == null) {
            this.yamlConfiguration.set(path.checkjstring(), Utilities.getObjectFromLuavalue(value));
            return LuaValue.TRUE;
        }

        return LuaValue.FALSE;
    }

    @Override
    public void setValue(LuaString path, LuaValue value) throws StorageObjectException {
        this.yamlConfiguration.set(path.checkjstring(), Utilities.getObjectFromLuavalue(value));
    }

    @Override
    protected LuaBoolean clearVaule(LuaString path) throws StorageObjectException {
        if (this.yamlConfiguration.get(path.checkjstring()) == null) {
            this.yamlConfiguration.set(path.checkjstring(), null);
            return LuaValue.TRUE;
        }

        return LuaValue.FALSE;
    }

    @Override
    public LuaValue getValue(LuaString path) throws StorageObjectException {
        return CoerceJavaToLua.coerce(this.yamlConfiguration.get(path.tojstring()));
    }

    @Override
    public void save() {
        this.preSaveCheck();
        try {
            this.yamlConfiguration.save(this.getStorageFile());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/io.thegamingmahi.luacord/environment/wrappers/storage/YamlStorage.kt
package io.thegamingmahi.luacord.environment.wrappers.storage

import io.thegamingmahi.luacord.Utilities
import io.thegamingmahi.luacord.environment.exception.StorageObjectException
import io.thegamingmahi.luacord.environment.plugin.LukkitPlugin
import org.bukkit.configuration.InvalidConfigurationException
import org.bukkit.configuration.file.YamlConfiguration
import org.luaj.vm2.LuaBoolean
import org.luaj.vm2.LuaString
import org.luaj.vm2.LuaValue
import org.luaj.vm2.lib.jse.CoerceJavaToLua
import java.io.FileReader
import java.io.IOException

/**
 * YAML storage implementation
 *
 * @author jammehcow
 * @author TheGamingMahi (Kotlin rewrite)
 */
class YamlStorage(plugin: LukkitPlugin, path: String) : StorageObject(plugin, path, Storage.YAML) {

    private val yamlConfiguration: YamlConfiguration = YamlConfiguration()

    init {
        try {
            yamlConfiguration.load(FileReader(getStorageFile()))
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: InvalidConfigurationException) {
            e.printStackTrace()
        }
    }

    override fun exists(path: String): LuaBoolean {
        return if (yamlConfiguration.get(path) != null) LuaValue.TRUE else LuaValue.FALSE
    }

    @Throws(StorageObjectException::class)
    override fun setDefaultValue(path: LuaString, value: LuaValue): LuaBoolean {
        return if (yamlConfiguration.get(path.checkjstring()) == null) {
            yamlConfiguration.set(path.checkjstring(), Utilities.getObjectFromLuavalue(value))
            LuaValue.TRUE
        } else {
            LuaValue.FALSE
        }
    }

    @Throws(StorageObjectException::class)
    override fun setValue(path: LuaString, value: LuaValue) {
        yamlConfiguration.set(path.checkjstring(), Utilities.getObjectFromLuavalue(value))
    }

    @Throws(StorageObjectException::class)
    override fun clearValue(path: LuaString): LuaBoolean {
        return if (yamlConfiguration.get(path.checkjstring()) != null) {
            yamlConfiguration.set(path.checkjstring(), null)
            LuaValue.TRUE
        } else {
            LuaValue.FALSE
        }
    }

    @Throws(StorageObjectException::class)
    override fun getValue(path: LuaString): LuaValue {
        return CoerceJavaToLua.coerce(yamlConfiguration.get(path.tojstring()))
    }

    override fun save() {
        preSaveCheck()
        try {
            yamlConfiguration.save(getStorageFile())
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}